umbraco/Umbraco-CMS
# V8 - Add Media Tracking of items added via Macro Parameters in RTE + Grid - issue #8131

### Prerequisites

- [ Y] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes https://github.com/umbraco/Umbraco-CMS/issues/8131

### Description

Currently, the new Media Tracking functionality that tracks when a Media item is in use on a particular page, and therefore 'whether it can be safely removed' does not take into consideration Media that might have been picked in a Macro Parameter and inserted into the Rich Text Editor or Grid.

This is quite a common scenario where developers may use a Macro to create an insertable Image Gallery / Responsive Image + Caption or neatly formatted List of PDF files.

It is the responsibility of the property editor to 'get references' that include Udis, but the 'Macro Picker' in the grid and RTE isn't really a property editor in its own right, and can have in-turn Macro Parameters with different property editors as the source, ~~in theory~~, when a Macro is inserted in we should loop through the Macro Parameters and call the GetReferences implementation for each type of Parameter... ~~however this presents a problem in that from the embedded markup of the macro in the rich text area or JSON of the grid we cannot easily infer the type of parameter without further 'look ups', which in this context might be expensive?~~

~~But **pragmatically** we 'DO' have the Macro Parameter Values, and the Single Media Picker Macro Parameter stores Udis, and so if a Macro Parameter Value contains a Media Udi, or a CSV of Media Udis, then this is all we need for Media Tracking purposes... (it doesn't matter how it is picked)~~

This PR adds a new **HtmlMacroParameterParser**, that is injected into RichTextPropertyEditor + GridPropertyEditor - and used in their GetReferences methods to parse either the HTML (or the GridControls) for Macros, and their Macro Parameter Values.

~~If the~~ Macro Parameter Values ~~are in media Udi format, they~~ are then returned as references for the page, and the Media item is 'tracked' as being used on the page.

~~Have added some 'tests' around some of the edge cases.~~

~~This will work for any Macro Parameter Editor that stores it's values as Udis or a CSV of Udis...~~

... fun fact... the Multiple Media Picker Macro Parameter, although it uses the same implementation as the Single Media Picker Macro Parameter - still stores it's picked items as integer ids - have created a separate PR here to address that: https://github.com/umbraco/Umbraco-CMS/pull/8388

## How to test this PR?

In the starter kit, create a Macro called Insert Image, that has a Macro Parameter of type Single Media Picker - and implement a Macro Partial View to write out the image, and allow this macro to be used in Rich Text Area and Grid.

Now in a Rich Text Area on a page, insert the Macro, and pick an image as the Macro Parameter.
Save and Publish
View the Media Item in the backoffice, and you should see this is now 'tracked' as being in use on this page.
In a different page that uses the grid, insert the Macro again, and pick the same or different image, and publish the page, visit the media item and you should find that is also being 'tracked' as being used on the grid page.

 Thanks for considering this contribution to Umbraco CMS!
